### PR TITLE
Feature/undo redo

### PR DESCRIPTION
**Pull request by shahanahmed86**
- listening to key-down events that'll trigger the undo or redo functionality.
- worked on some optimization like using memo/callback where needed.
- made update/remove callback on the Todo component and remove from deep down the tree to implement the undo/redo functionality

### CODE REVIEW

**Comment by kamalkishor1991** [comment]
@shahanahmed86 thanks a lot for the pull request. I will review it in few days.

**Comment by shahanahmed86** [comment]
> @shahanahmed86 thanks a lot for the pull request. I will review it in few days.

This is my first contribution to the open-source community and I am not an expert on team collaboration and GITOPS things. I found anything that I need to work on then please suggest me.

**Review by kamalkishor1991 — CHANGES_REQUESTED** [review]

File: src/components/Todo/Todo.tsx [hunk 1]
```diff
@@ -11,15 +11,150 @@ export interface TodoAppProps {
...
+        setItemsCallback(prevItems);
+
+        break;
+      }
+		}
+	}, [redoItems, items, setItemsCallback]);
+
+  useEffect(() => {
+    const onUndoOrRedo = ({ repeat, ctrlKey, key }: KeyboardEvent) => {
+      if (repeat || !ctrlKey) return;
```
[the remark this answers is not in the document]
Comment (shahanahmed86) [reply src/components/Todo/Todo.tsx]: I have visited the shared [link](https://stackoverflow.com/questions/13319723/listen-to-undo-redo-event-in-contenteditable-div) but the [InputEvent](https://w3c.github.io/input-events/#interface-InputEvent-Attributes) was not properly working with react, it triggers the callback _too many times_ on a single key press. Instead, I've done my research on the `KeyboardEvent` and found that the `metaKey` `boolean` property gets true when next to the spacebar:
- [x] <kbd>Command</kbd> is pressed on Macbook
- [x] <kbd>Window</kbd> is pressed on Linux/Windows
And I've tested it on Macbook it works like a charm. I have updated this pull request also

[the remark this answers is not in the document]
Comment (shahanahmed86) [reply src/components/Todo/Todo.tsx]: @kamalkishor1991 but the problem is that <kbd>⌘y</kbd> is pressed it also opens the browser history and this is the browser behavior.

[the remark this answers is not in the document]
Comment (kamalkishor1991) [reply src/components/Todo/Todo.tsx]: @shahanahmed86 I think you should call `event.preventDefault();`
I tried google keep and it is able to handle cmd + y without opening browser history.

**Review by kamalkishor1991 — CHANGES_REQUESTED** [review]

File: src/components/Todo/Todo.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (shahanahmed86) [reply src/components/Todo/Todo.tsx]: @kamalkishor1991, I've updated the pull request

File: src/components/Todo/common/Todo/Form/index.tsx [hunk 2]
```diff
@@ -47,6 +60,14 @@ export const Form = (props: AddProps) => {
       <FormControl fullWidth>
         <TextField
           inputRef={inputRef}
+          InputProps={{
+            endAdornment: (
+              <>
+                <UndoIcon onClick={onUndo} className={classes.iconStyle} />
```
[the remark this answers is not in the document]
Comment (shahanahmed86) [reply src/components/Todo/common/Todo/Form/index.tsx:66]: got it

[the remark this answers is not in the document]
Comment (kamalkishor1991) [reply src/components/Todo/common/Todo/Form/index.tsx:66]: No not this header. this is storybook header. I meant that we can introduce a action bar on top of input box and add the icons in the action bar.

[the remark this answers is not in the document]
Comment (shahanahmed86) [reply src/components/Todo/common/Todo/Form/index.tsx]: I've updated the pull request with toolbar shortcut implementation

[the remark this answers is not in the document]
Comment (shahanahmed86) [reply src/components/Todo/common/Todo/Form/index.tsx]: @kamalkishor1991 I've found some conflicts with the master that I was trying to merge. I assume resolving conflicts is, for the person, who creates/updates the pull request. So I am working on that.

[the remark this answers is not in the document]
Comment (kamalkishor1991) [reply src/components/Todo/common/Todo/Form/index.tsx]: yes I merged one PR( #53) yesterday. You can also take a look at the PR to resolve the conflicts correctly.

**Comment by shahanahmed86** [comment]
@kamalkishor1991, I've updated the pull request with all conflicts resolved and deployed the app on this [URL](https://shahanahmed86.github.io/react-todo). Please check the demo also before merging.

**Comment by kamalkishor1991** [comment]
I found 2 issues.
1. The action bar width should match 
Screenshot 2023-05-28 at 1 44 41 PM
the bottom area
2. When adding more items the add item is merging with action bar.
Please check

**Comment by shahanahmed86** [comment]
@kamalkishor1991, The changes that you ask to fix are now updated in the last commit:

- making toolbar width according to the Todo component.
- when adding more items the text field is merging in the toolbar.

**Review by kamalkishor1991 — APPROVED** [review]

**Pull request merged**